### ISSUE TITLE
Which function to override if one wants to clip_grad_norm on some, not all, parameters?

### ISSUE THREAD

**Issue by LeeJZh** [PyTorchLightning/pytorch-lightning#5982]
### Before asking:
1. Try to find answers to your questions in [the Lightning Forum!](https://forums.pytorchlightning.ai/)
2. Search for similar [issues](https://github.com/PyTorchLightning/pytorch-lightning/issues).
3. Search the [docs](https://pytorch-lightning.readthedocs.io/en/latest/).

#### What is your question?
Which function to override if one wants to clip_grad_norm on some, not all, parameters?

#### What's your environment?

 - OS: [e.g. iOS, Linux, Win]
 - Packaging [e.g. pip, conda]
 - Version [e.g. 0.5.2.1]

**Comment by tchaton** [comment]
Hey there,

Currently in master, the `clip_gradients` lives in the precision plugin.

```
from pytorch_lightning.plugins.precision import NativeAMPPlugin

class MyCustomAMPPlugin(NativeAMPPlugin):

    def clip_gradients(self, grad_clip_val: Union[int, float], optimizer: Optimizer, norm_type: float):
        # Example: clip gradients only on the model head
        model = self.trainer.get_model()
        torch.nn.utils.clip_grad_norm_(model.head.parameters(), max_norm=grad_clip_val, norm_type=norm_type)

Trainer(amp_backend='native', precision=16, plugins=[MyCustomAMPPlugin()])
```

**Comment by SeanNaren** [comment]
Related to this: https://github.com/PyTorchLightning/pytorch-lightning/issues/3912

Having `clip_gradients` as a part of the module makes sense till we realise that different training type/accelerators do different things when clipping gradient norms based on precision. Before this used to lead to us having a bunch of if statements per accelerator in the function within the lightning module, but I think that's not ideal.

I think as you suggested @tchaton is the best way to go, but I might be missing something. In the future I think clip_gradients should live with the training type plugin, but still be overridden as you suggested

**Comment by LeeJZh** [comment]
> Hey there,
> 
> Currently in master, the `clip_gradients` lives in the precision plugin.
> 
> ```
> from pytorch_lightning.plugins.precision import NativeAMPPlugin
> 
> class MyCustomAMPPlugin(NativeAMPPlugin):
> 
>     def clip_gradients(self, grad_clip_val: Union[int, float], optimizer: Optimizer, norm_type: float):
>         # Example: clip gradients only on the model head
>         model = self.trainer.get_model()
>         torch.nn.utils.clip_grad_norm_(model.head.parameters(), max_norm=grad_clip_val, norm_type=norm_type)
> 
> Trainer(amp_backend='native', precision=16, plugins=[MyCustomAMPPlugin()])
> ```
> 
> However, this would have to be done for all precision plugin.
> 
> We are currently thinking about providing a hook to enable more flexibility there.
> 
> @SeanNaren Any idea ?
> 
> Best,
> T.C

Thanks, that helps a lot!

**Comment by LeeJZh** [comment]
> Hey there,
> 
> Currently in master, the `clip_gradients` lives in the precision plugin.
> 
> ```
> from pytorch_lightning.plugins.precision import NativeAMPPlugin
> 
> class MyCustomAMPPlugin(NativeAMPPlugin):
> 
>     def clip_gradients(self, grad_clip_val: Union[int, float], optimizer: Optimizer, norm_type: float):
>         # Example: clip gradients only on the model head
>         model = self.trainer.get_model()
>         torch.nn.utils.clip_grad_norm_(model.head.parameters(), max_norm=grad_clip_val, norm_type=norm_type)
> 
> Trainer(amp_backend='native', precision=16, plugins=[MyCustomAMPPlugin()])
> ```
> 
> However, this would have to be done for all precision plugin.
> 
> We are currently thinking about providing a hook to enable more flexibility there.
> 
> @SeanNaren Any idea ?
> 
> Best,
> T.C

Hi, I keep getting 

> 'NoneType' object has no attribute 'scaler'

error

with 
`
trainer = pl.Trainer.from_argparse_args(args,
                                            logger=logger,
                                            callbacks=[checkpoint_callback], plugins=[MyCustomAMPPlugin()])
`
i guess i fail to pass tranier to plugins?
